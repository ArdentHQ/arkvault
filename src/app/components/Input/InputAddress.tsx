import { Contracts } from "@ardenthq/sdk-profiles";
import React from "react";
import { RegisterOptions } from "react-hook-form";
import { useTranslation } from "react-i18next";

import { Input } from "./Input";

export type InputAddressProperties = {
	profile: Contracts.IProfile;
	coin?: string;
	network?: string;
	registerRef?: (options: RegisterOptions) => (reference: HTMLInputElement | null) => void;
	additionalRules?: RegisterOptions;
	onValidAddress?: (address: string) => void;
	onChange?: (address: string) => void;
	useDefaultRules?: boolean;
} & React.InputHTMLAttributes<any>;

const defaultProps = {
	additionalRules: {},
};

export const InputAddress = ({
	profile,
	coin,
	network,
	registerRef,
	additionalRules = defaultProps.additionalRules,
	onValidAddress,
	useDefaultRules = true,
	...properties
}: InputAddressProperties) => {
	const { t } = useTranslation();

	const validateAddress = async (address: string) => {
		const instance = profile.coins().set(coin!, network!);
		const isValidAddress: boolean = await instance.address().validate(address);

		if (isValidAddress) {
			onValidAddress?.(address);
			return true;
		}

		return t("COMMON.INPUT_ADDRESS.VALIDATION.NOT_VALID");
	};

	const defaultRules = {
		...additionalRules,
		validate: {
			validateAddress,
			...additionalRules?.validate,
		},
	};
	const rules = useDefaultRules ? defaultRules : {};

	return (
		<Input
			ref={registerRef?.(rules)}
			type="text"
			data-testid="InputAddress__input"
			{...properties}
			autoComplete="off"
		/>
	);
};
