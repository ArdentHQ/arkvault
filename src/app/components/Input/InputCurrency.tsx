import { Currency } from "@payvo/sdk-intl";
import React, { useEffect, useState } from "react";

import { Input } from "./Input";

type InputCurrencyProperties = {
	addons?: any;
	onChange?: (value: any) => void;
	innerClassName?: string;
	ignoreContext?: boolean;
	errorMessage?: string;
	isInvalid?: boolean;
	isCompact?: boolean;
	noShadow?: boolean;
} & Omit<React.InputHTMLAttributes<any>, "onChange" | "defaultValue">;

const sanitize = (value?: string) => Currency.fromString(value || "").display;

export const InputCurrency = React.forwardRef<HTMLInputElement, InputCurrencyProperties>(
	({ onChange, value, ...properties }: InputCurrencyProperties, reference) => {
		const [amount, setAmount] = useState<string>(sanitize(value?.toString()));

		useEffect(() => {
			// when value is changed outside, update amount as well
			setAmount(sanitize(value?.toString()));
		}, [value]);

		const handleInput = (event: React.ChangeEvent<HTMLInputElement>) => {
			const sanitizedValue = sanitize(event.target.value);

			setAmount(sanitizedValue);
			onChange?.(sanitizedValue);
		};

		return (
			<div className="relative">
				<Input
					data-testid="InputCurrency"
					onChange={handleInput}
					ref={reference}
					type="text"
					value={amount}
					{...properties}
				/>
			</div>
		);
	},
);

InputCurrency.displayName = "InputCurrency";
