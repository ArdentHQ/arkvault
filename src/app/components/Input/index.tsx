export * from "./Input";
export * from "./InputAddress";
export * from "./InputCounter";
export * from "./InputCurrency";
export * from "./InputDate";
export * from "./InputDefault";
export * from "./InputPassword";
