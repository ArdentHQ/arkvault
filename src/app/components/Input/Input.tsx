import cn from "classnames";
import React, { useEffect, useRef } from "react";
import tw, { styled } from "twin.macro";

import { InputSuggestion } from "./InputSuggestion";
import { useFormField } from "@/app/components/Form/useFormField";
import { Icon } from "@/app/components/Icon";
import { Tooltip } from "@/app/components/Tooltip";

interface AddonProperties {
	wrapperClassName?: string;
	content: JSX.Element;
}

type InputProperties = {
	addons?: {
		start?: AddonProperties;
		end?: AddonProperties;
	};
	as?: React.ElementType;
	errorMessage?: string;
	hideInputValue?: boolean;
	ignoreContext?: boolean;
	innerClassName?: string;
	isFocused?: boolean;
	isInvalid?: boolean;
	isTextArea?: boolean;
	isCompact?: boolean;
	noBorder?: boolean;
	noShadow?: boolean;
	suggestion?: string;
} & React.HTMLProps<any>;

export const InputWrapperStyled = styled.div<{
	disabled?: boolean;
	invalid?: boolean;
	isTextArea?: boolean;
	isCompact?: boolean;
	noBorder?: boolean;
	noShadow?: boolean;
}>`
	${tw`flex items-center w-full px-4 space-x-2 transition-colors duration-200 rounded appearance-none text-theme-text`}

	${({ noBorder }) => {
		if (!noBorder) {
			return tw`border`;
		}
	}}

	${({ noShadow }) => {
		if (!noShadow) {
			return tw`focus-within:ring-1`;
		}
	}}

	${({ disabled, invalid }) => {
		if (disabled && invalid) {
			return tw`border-theme-danger-500 bg-theme-secondary-100 dark:bg-theme-secondary-800`;
		}

		if (disabled) {
			return tw`border-theme-secondary-300 dark:border-theme-secondary-700 bg-theme-secondary-100 dark:bg-theme-secondary-800`;
		}

		if (invalid) {
			return tw`bg-theme-background border-theme-danger-500 focus-within:ring-theme-danger-500`;
		}

		return tw`bg-theme-background border-theme-secondary-400 dark:border-theme-secondary-700 focus-within:(border-theme-primary-600 ring-theme-primary-600)`;
	}}

	${({ isTextArea, isCompact }) => {
		if (isTextArea) {
			return tw`relative`;
		}

		if (isCompact) {
			return tw`height[34px] overflow-hidden`;
		}

		return tw`h-14 overflow-hidden`;
	}}
`;

const InputStyled = styled.input`
	${tw`!bg-transparent !p-0`}

	&:focus {
		${tw`outline-none shadow-none (ring-0 ring-transparent)!`}
	}
	&::placeholder {
		${tw`text-theme-secondary-400 dark:text-theme-secondary-700`}
	}
	&.shadow-none {
		${tw`shadow-none`}
	}
`;

type InputElement = HTMLInputElement | HTMLSelectElement | HTMLTextAreaElement;

export const Input = React.forwardRef<InputElement, InputProperties>(
	(
		{
			addons,
			className,
			disabled,
			errorMessage,
			hideInputValue,
			ignoreContext,
			innerClassName,
			isFocused,
			isInvalid,
			isTextArea,
			isCompact,
			noBorder,
			noShadow,
			style,
			suggestion,
			value,
			...properties
		}: InputProperties,
		reference,
	) => {
		let fieldContext = useFormField();

		if (ignoreContext) {
			fieldContext = undefined;
		}

		const isInvalidValue = fieldContext?.isInvalid || isInvalid;
		const errorMessageValue = fieldContext?.errorMessage || errorMessage;

		const focusReference = useRef<InputElement>(null);

		reference = isFocused ? focusReference : reference;

		useEffect(() => {
			if (isFocused && focusReference.current) {
				focusReference.current.focus();
			}
		}, [focusReference, isFocused]);

		const hiddenReference = useRef<HTMLDivElement>(null);

		return (
			<>
				{suggestion && (
					<div ref={hiddenReference} className="invisible fixed w-auto whitespace-nowrap">
						{value}â€¦
					</div>
				)}

				<InputWrapperStyled
					style={style}
					className={className}
					disabled={disabled}
					invalid={isInvalidValue}
					noBorder={noBorder}
					noShadow={noShadow}
					isTextArea={isTextArea}
					isCompact={isCompact}
				>
					{addons?.start !== undefined && addons.start.content}
					<div className={cn("relative flex h-full flex-1", { invisible: hideInputValue })}>
						<InputStyled
							data-testid="Input"
							className={cn("no-ligatures w-full border-none", innerClassName, {
								"text-theme-secondary-text": disabled,
							})}
							name={fieldContext?.name}
							aria-invalid={isInvalidValue}
							disabled={disabled}
							value={value}
							type="text"
							// @ts-ignore
							ref={reference}
							{...properties}
							autoComplete="off"
						/>

						<InputSuggestion
							suggestion={suggestion}
							hiddenReference={hiddenReference}
							innerClassName={innerClassName}
						/>
					</div>

					{(isInvalidValue || addons?.end) && (
						<div
							data-testid="Input__addon-end"
							className={cn(
								"flex items-center space-x-3 divide-x divide-theme-secondary-300 dark:divide-theme-secondary-800",
								{
									"absolute bottom-full right-0 mb-2": isTextArea,
									"text-theme-danger-500": isInvalidValue,
									"text-theme-primary-300 dark:text-theme-secondary-600": !isInvalidValue,
								},
								addons?.end?.wrapperClassName,
							)}
						>
							{isInvalidValue && (
								<Tooltip content={errorMessageValue} size="sm">
									<span data-errortext={errorMessageValue} data-testid="Input__error">
										<Icon
											name="CircleExclamationMark"
											className="text-theme-danger-500"
											size="lg"
										/>
									</span>
								</Tooltip>
							)}

							{addons?.end && (
								<div className={cn({ "pl-3": isInvalidValue && !addons.end.wrapperClassName })}>
									{addons.end.content}
								</div>
							)}
						</div>
					)}
				</InputWrapperStyled>
			</>
		);
	},
);

Input.displayName = "Input";
