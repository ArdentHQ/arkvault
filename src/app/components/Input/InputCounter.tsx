import cn from "classnames";
import React, { useState } from "react";

import { useFormField } from "@/app/components/Form/useFormField";
import { Input } from "@/app/components/Input";

type Properties = {
	maxLength?: number;
	maxLengthLabel?: string;
	defaultValue?: string;
} & React.InputHTMLAttributes<any>;

export const InputCounter = React.forwardRef<HTMLInputElement, Properties>((properties: Properties, reference) => {
	const fieldContext = useFormField();
	const [length, setLength] = useState(properties.defaultValue?.length || 0);

	const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		setLength(event.target.value.length);
		properties.onChange?.(event);
	};

	return (
		<Input
			data-testid="InputCounter__input"
			ref={reference}
			{...properties}
			onChange={handleChange}
			addons={{
				end: {
					content: (
						<span
							data-testid="InputCounter__counter"
							className={cn("text-sm font-semibold", {
								"text-theme-secondary-500 dark:text-theme-secondary-700": !fieldContext?.isInvalid,
							})}
						>
							{length}/{properties.maxLengthLabel}
						</span>
					),
				},
			}}
		/>
	);
});

InputCounter.displayName = "InputCounter";
