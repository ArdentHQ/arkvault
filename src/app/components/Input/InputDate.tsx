import React from "react";
import DatePicker from "react-datepicker";

import { Controller, useFormContext } from "react-hook-form";
import { Input } from "./Input";
import { Icon } from "@/app/components/Icon";
import { useFormField } from "@/app/components/Form/useFormField";

type InputDateProperties = {
	minDate?: Date;
	startDate?: Date;
	placement?: string;
	selectsStart?: boolean;
	selectsEnd?: boolean;
	isInvalid?: boolean;
} & React.InputHTMLAttributes<any>;

export const InputDate = React.forwardRef<HTMLInputElement, InputDateProperties>(
	({ minDate, startDate, placement = "bottom-start", selectsStart, selectsEnd, ...properties }, reference) => {
		const { control } = useFormContext();

		const fieldContext = useFormField();

		return (
			<Controller
				name={fieldContext!.name}
				control={control}
				render={(field) => (
					<DatePicker
						selected={field.value}
						calendarClassName="bg-theme-background"
						popperPlacement={placement as any}
						selectsStart={selectsStart}
						selectsEnd={selectsEnd}
						minDate={minDate}
						startDate={startDate}
						onChange={field.onChange}
						customInput={
							<Input
								data-testid="InputDate"
								ref={reference}
								addons={{
									end: {
										content: (
											<button
												data-testid="InputDate__calendar"
												type="button"
												// onClick={() => ???.setOpen(true)}
												className="ring-focus relative flex h-full w-full items-center justify-center text-2xl focus:outline-none"
												data-ring-focus-margin="-m-1"
											>
												<Icon name="Calendar" size="lg" />
											</button>
										),
									},
								}}
								isInvalid={!field.value}
								{...properties}
							/>
						}
					/>
				)}
			/>
		);
	},
);

InputDate.displayName = "InputDate";
