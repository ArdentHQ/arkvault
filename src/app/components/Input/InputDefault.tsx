import cn from "classnames";
import React from "react";

import { Input } from "@/app/components/Input";

type InputDefaultProperties = {
	className?: string;
} & React.InputHTMLAttributes<any>;

export const InputDefault = React.forwardRef<HTMLInputElement, InputDefaultProperties>(
	({ className, ...properties }: InputDefaultProperties, reference) => (
		<Input className={cn(className)} ref={reference} {...properties} />
	),
);

InputDefault.displayName = "InputDefault";
