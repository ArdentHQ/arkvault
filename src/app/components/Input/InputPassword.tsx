import React from "react";

import { Input } from "./Input";
import { Icon } from "@/app/components/Icon";

type InputPasswordProperties = { isInvalid?: boolean } & React.InputHTMLAttributes<any>;

export const InputPassword = React.forwardRef<HTMLInputElement, InputPasswordProperties>((properties, reference) => {
	const [show, setShow] = React.useState(false);
	const togglePasswordVisibility = () => setShow(!show);

	return (
		<Input
			data-testid="InputPassword"
			ref={reference}
			type={show ? "text" : "password"}
			autoComplete="new-password"
			addons={{
				end: {
					content: (
						<button
							data-testid="InputPassword__toggle"
							type="button"
							onClick={togglePasswordVisibility}
							className="ring-focus relative flex h-full w-full items-center justify-center text-2xl focus:outline-none"
							data-ring-focus-margin="-m-1"
						>
							<Icon name={show ? "EyeSlash" : "Eye"} size="lg" />
						</button>
					),
				},
			}}
			{...properties}
		/>
	);
});

InputPassword.displayName = "InputPassword";
