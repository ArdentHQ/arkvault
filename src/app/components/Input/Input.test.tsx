import React from "react";

import { Input } from "./Input";
import { render, screen } from "@/utils/testing-library";

describe("Input", () => {
	it.each([true, false])("should render when isCompact = %s", (isCompact: boolean) => {
		const { asFragment } = render(<Input isCompact={isCompact} />);

		expect(asFragment()).toMatchSnapshot();
	});

	it("should render as custom element", () => {
		const { asFragment } = render(<Input as="select" />);
		const input = screen.getByTestId("Input");

		expect(input.tagName).toBe("SELECT");

		expect(asFragment()).toMatchSnapshot();
	});

	it("should render with focus", () => {
		const { asFragment } = render(<Input isFocused />);

		expect(asFragment()).toMatchSnapshot();
	});

	it("should render with invalid", () => {
		const { asFragment } = render(<Input isInvalid={true} errorMessage="Field invalid" />);
		const input = screen.getByTestId("Input__error");

		expect(input).toBeVisible();

		expect(asFragment()).toMatchSnapshot();
	});

	it("should render with addons", () => {
		const { asFragment } = render(
			<Input
				addons={{
					end: {
						content: <span>end</span>,
						wrapperClassName: "some-class-name",
					},
					start: {
						content: <span>start</span>,
					},
				}}
			/>,
		);

		const addonWrapper = screen.getByTestId("Input__addon-end");

		expect(addonWrapper).toBeInTheDocument();
		expect(addonWrapper).toHaveClass("some-class-name");
		expect(addonWrapper).toHaveTextContent("end");

		expect(screen.getByText("start")).toBeInTheDocument();

		expect(asFragment()).toMatchSnapshot();
	});

	it("should render with suggestion", () => {
		const { asFragment } = render(<Input suggestion="suggestion" />);

		expect(screen.getByTestId("Input__suggestion")).toBeInTheDocument();

		expect(asFragment()).toMatchSnapshot();
	});
});
