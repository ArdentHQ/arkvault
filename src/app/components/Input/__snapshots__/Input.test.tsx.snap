// Vitest Snapshot v1

exports[`Input > should render as custom element 1`] = `
<DocumentFragment>
  <div
    class="css-1psxlcd"
  >
    <div
      class="relative flex h-full flex-1"
    >
      <select
        autocomplete="off"
        class="no-ligatures w-full border-none css-1nupw50"
        data-testid="Input"
        type="text"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input > should render as textarea 1`] = `
<DocumentFragment>
  <div
    class="css-15amhun"
  >
    <div
      class="relative flex h-full flex-1"
    >
      <input
        autocomplete="off"
        class="no-ligatures w-full border-none css-1nupw50"
        data-testid="Input"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input > should render disabled 1`] = `
<DocumentFragment>
  <div
    class="css-n29l73"
    disabled=""
  >
    <div
      class="relative flex h-full flex-1"
    >
      <input
        autocomplete="off"
        class="no-ligatures w-full border-none text-theme-secondary-text css-1nupw50"
        data-testid="Input"
        disabled=""
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input > should render invalid 1`] = `
<DocumentFragment>
  <div
    class="css-px3dod"
  >
    <div
      class="relative flex h-full flex-1"
    >
      <input
        aria-invalid="true"
        autocomplete="off"
        class="no-ligatures w-full border-none css-1nupw50"
        data-testid="Input"
        type="text"
        value=""
      />
    </div>
    <div
      class="flex items-center space-x-3 divide-x divide-theme-secondary-300 dark:divide-theme-secondary-800 text-theme-danger-500"
      data-testid="Input__addon-end"
    >
      <span
        data-errortext="Field invalid"
        data-testid="Input__error"
      >
        <div
          class="text-theme-danger-500 css-1i4b0eq"
          height="20"
          width="20"
        >
          <svg
            id="circle-exclamation-mark"
            viewBox="0 0 20 20"
            x="0"
            xml:space="preserve"
            xmlns="http://www.w3.org/2000/svg"
            y="0"
          >
            <g
              fill="none"
              stroke="currentColor"
              stroke-width="2"
            >
              <circle
                cx="10"
                cy="10"
                r="9"
                stroke-miterlimit="10"
              />
              <path
                d="M10 10.5v-5"
                stroke-linecap="round"
                stroke-miterlimit="10"
              />
              <path
                d="M10 14h0"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </g>
          </svg>
        </div>
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input > should render invalid and disabled 1`] = `
<DocumentFragment>
  <div
    class="css-1by871c"
    disabled=""
  >
    <div
      class="relative flex h-full flex-1"
    >
      <input
        aria-invalid="true"
        autocomplete="off"
        class="no-ligatures w-full border-none text-theme-secondary-text css-1nupw50"
        data-testid="Input"
        disabled=""
        type="text"
        value=""
      />
    </div>
    <div
      class="flex items-center space-x-3 divide-x divide-theme-secondary-300 dark:divide-theme-secondary-800 text-theme-danger-500"
      data-testid="Input__addon-end"
    >
      <span
        data-errortext="Field invalid and disabled"
        data-testid="Input__error"
      >
        <div
          class="text-theme-danger-500 css-1i4b0eq"
          height="20"
          width="20"
        >
          <svg
            id="circle-exclamation-mark"
            viewBox="0 0 20 20"
            x="0"
            xml:space="preserve"
            xmlns="http://www.w3.org/2000/svg"
            y="0"
          >
            <g
              fill="none"
              stroke="currentColor"
              stroke-width="2"
            >
              <circle
                cx="10"
                cy="10"
                r="9"
                stroke-miterlimit="10"
              />
              <path
                d="M10 10.5v-5"
                stroke-linecap="round"
                stroke-miterlimit="10"
              />
              <path
                d="M10 14h0"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </g>
          </svg>
        </div>
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input > should render when isCompact = false 1`] = `
<DocumentFragment>
  <div
    class="css-1psxlcd"
  >
    <div
      class="relative flex h-full flex-1"
    >
      <input
        autocomplete="off"
        class="no-ligatures w-full border-none css-1nupw50"
        data-testid="Input"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input > should render when isCompact = true 1`] = `
<DocumentFragment>
  <div
    class="css-6zwcjo"
  >
    <div
      class="relative flex h-full flex-1"
    >
      <input
        autocomplete="off"
        class="no-ligatures w-full border-none css-1nupw50"
        data-testid="Input"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input > should render with addons 1`] = `
<DocumentFragment>
  <div
    class="css-1psxlcd"
  >
    <span>
      start
    </span>
    <div
      class="relative flex h-full flex-1"
    >
      <input
        autocomplete="off"
        class="no-ligatures w-full border-none css-1nupw50"
        data-testid="Input"
        type="text"
        value=""
      />
    </div>
    <div
      class="flex items-center space-x-3 divide-x divide-theme-secondary-300 dark:divide-theme-secondary-800 text-theme-primary-300 dark:text-theme-secondary-600 some-class-name"
      data-testid="Input__addon-end"
    >
      <div
        class=""
      >
        <span>
          end
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input > should render with focus 1`] = `
<DocumentFragment>
  <div
    class="css-1psxlcd"
  >
    <div
      class="relative flex h-full flex-1"
    >
      <input
        autocomplete="off"
        class="no-ligatures w-full border-none css-1nupw50"
        data-testid="Input"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input > should render with suggestion 1`] = `
<DocumentFragment>
  <div
    class="invisible fixed w-auto whitespace-nowrap"
  >
    â€¦
  </div>
  <div
    class="css-1psxlcd"
  >
    <div
      class="relative flex h-full flex-1"
    >
      <input
        autocomplete="off"
        class="no-ligatures w-full border-none css-1nupw50"
        data-testid="Input"
        type="text"
        value=""
      />
      <span
        class="pointer-events-none absolute inset-y-0 flex w-full items-center font-normal opacity-50"
        data-testid="Input__suggestion"
      >
        <span
          class="truncate"
        >
          suggestion
        </span>
      </span>
    </div>
  </div>
</DocumentFragment>
`;
